### PR TITLE
Update HostFactoryResolver to account for lack of application name with new host pattern

### PR DESCRIPTION
**Pull request by ajcvickers**
Based on https://github.com/dotnet/efcore/pull/26198

The fix should go here, rather than in EF Core, because then anybody using this code will get the fixed behavior. I have tested this locally and it fixes https://github.com/dotnet/efcore/issues/26177. However tests should be added to this repo.

### CODE REVIEW

**Comment by ajcvickers** [comment]
@davidfowl You're probably the best person to add tests here.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs [hunk 1]
```diff
@@ -144,6 +145,14 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
             {
                 return args =>
                 {
+                    static bool IsApplicationNameArg(string arg)
+                        => arg.Equals("--applicationName", StringComparison.OrdinalIgnoreCase) ||
+                            arg.Equals("/applicationName", StringComparison.OrdinalIgnoreCase);
+
+                    args = args.Any(arg => IsApplicationNameArg(arg)) || assembly.FullName is null
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:152]: When is `assembly.FullName` null?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:152]: When you're writing unit tests 😄

**Comment by maryamariyan** [comment]
@ajcvickers thanks for initiating the fix here. Seems like it needs unit test to go with it. Are you planning on adding the test for this? If you'd prefer, I could take over the PR from here, let me know thanks.

**Comment by ajcvickers** [comment]
@maryamariyan I was hoping @davidfowl would do it, since he made the change that broke things. Or am I expecting too much from a partner architect? :trollface: But if you want to let him off easy then go for it...

**Comment by davidfowl** [comment]
> Or am I expecting too much from a partner architect? :trollface: But if you want to let him off easy then go for it...

Entirely, but we have a slew of wonderful tests here https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/HostFactoryResolverTests.cs 😄

**Review by bricelam — APPROVED** [review]

**Pull request closed without merging**